### PR TITLE
Add BigQuery as a connectivity type to use in sql actions

### PR DESCRIPTION
**Pull request by peterbillen**
**Is your feature request related to a problem? Please describe.**
I would like to allow iesi scripts to use its sql actions on BigQuery

**Describe the solution you'd like**
Add BigQuery as a connection type in iesi using the jdbc driver

**Describe alternatives you've considered**
NA

**Additional context**
NA

**ID**
473202d

### CODE REVIEW

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/DatabaseService.java [hunk 1]
```diff
@@ -26,6 +26,11 @@
 public abstract class DatabaseService<T extends Database> implements IDatabaseService<T> {
 
     public Connection getConnection(T database) throws SQLException {
+        //Avoid creating a Hikari Connection pool for Bigquery
+        if (database.getDatabaseConnection().getType().equalsIgnoreCase("bigquery")) {
+            return DatabaseConnectionHandler.getInstance().getConnection(database.getDatabaseConnection());
+        }
+
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/DatabaseService.java:33]: Method 'public Connection getConnection(T database) throws SQLException' should be overriden in 'core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/BigqueryDatabaseService.java'. See 'io.metadew.iesi.connection.database.sqlite.SqliteDatabaseService' for example

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/connection/DatabaseConnectionService.java [hunk 2]
```diff
@@ -17,8 +17,14 @@
...
-            Connection connection = DriverManager.getConnection(databaseConnection.getConnectionURL(), databaseConnection.getUserName(), databaseConnection.getUserPassword());
-            connection.setAutoCommit(false);
+            Connection connection;
+            //Ensure Bigquery connection url can be used
+            if (databaseConnection.getType().equalsIgnoreCase("bigquery")) {
+                connection = DriverManager.getConnection(databaseConnection.getConnectionURL());
+            } else {
+                connection = DriverManager.getConnection(databaseConnection.getConnectionURL(), databaseConnection.getUserName(), databaseConnection.getUserPassword());
+                connection.setAutoCommit(false);
+            }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/connection/DatabaseConnectionService.java:27]: Method 'public Connection getConnection(T databaseConnection)' should be overriden in 'core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/BigqueryDatabaseConnectionService.java'

File: core/java/iesi-core/src/main/resources/log4j2.xml [hunk 3]
```diff
@@ -139,7 +139,7 @@
             <AppenderRef ref="queries" level="all" />
         </Logger>
         <Logger name="io.metadew.iesi.script" additivity="false">
-            <AppenderRef ref="STDOUT" level="info" />
+            <AppenderRef ref="STDOUT" level="debug" />
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/resources/log4j2.xml:142]: Why is this changed to debug?

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/BigqueryDatabaseService.java [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+                                DatabaseHandler.getInstance().getMandatoryParameterWithKey(connection, clientSecretKey)));
+                return new BigqueryDatabase(bigqueryDatabaseConnection, schemaName);
+            case 3:
+                bigqueryDatabaseConnection = new BigqueryDatabaseConnection(
+                        getConnectionUrlDefaultCredentials(
+                                host,
+                                port,
+                                project,
+                                authValue));
+                return new BigqueryDatabase(bigqueryDatabaseConnection, schemaName);
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/BigqueryDatabaseService.java:94]: Create 4 subclasses of BigqueryDatabaseConnection for each auth type. Constructor of each subclass only receives the necessary arguments needed for the specific auth type. See 'io.metadew.iesi.connection.database.oracle.OracleDatabaseService' and 'io.metadew.iesi.connection.database.oracle.ServiceNameOracleDatabaseConnection' for example

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/BigqueryDatabaseService.java [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+            connectionUrl.append(portNumber);
+        }
+        connectionUrl.append(";ProjectId=");
+        connectionUrl.append(project);
+        connectionUrl.append(";OAuthType=");
+        connectionUrl.append(authValue);
+        connectionUrl.append(";");
+
+        return connectionUrl.toString();
+    }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/BigqueryDatabaseService.java:186]: Should be deleted/moved once subclasses of 'BigqueryDatabaseConnection' are implemented

File: docs/pages/define/connectiontypes/db.bigquery.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+* define the name of the dataset
+
+### 4: authValue
+
+`authValue: "authValue number"`
+* define the number that specifies the type of authentication used by the driver
+  * 0: The driver uses service-based OAuth authentication
+  * 1: The driver uses user-based OAuth authentication
+  * 2: The driver uses pre-generated tokens for authentication
+  * 3: The driver uses Application Default Credentials for authentication
```
Comment (BerrevoetsRobbe) [inline docs/pages/define/connectiontypes/db.bigquery.md:44]: maybe rename to authMode and use more functional options like 'service', 'user', 'token' and 'application default credentials' instead of 0, 1, 2, 3. This keeps it more in line with Oracle as well, cfr. 'mode' and 'tnsAlias'/'service'

**Review by BerrevoetsRobbe — CHANGES_REQUESTED** [review]

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/DatabaseService.java (hunk 1 above)
Comment (peterbillen) [reply core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/DatabaseService.java:33]: Resolved in next commit

File: core/java/iesi-core/src/main/resources/log4j2.xml (hunk 3 above)
Comment (peterbillen) [reply core/java/iesi-core/src/main/resources/log4j2.xml:142]: correct, temporarily set for debugging issues, will be set correctly in next commit

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/BigqueryDatabaseService.java (hunk 5 above)
Comment (peterbillen) [reply core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/BigqueryDatabaseService.java:186]: Resolve in next commit

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/BigqueryDatabaseService.java (hunk 4 above)
Comment (peterbillen) [reply core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/BigqueryDatabaseService.java:94]: Resolved in next commit

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/connection/DatabaseConnectionService.java (hunk 2 above)
Comment (peterbillen) [reply core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/connection/DatabaseConnectionService.java:27]: Resolved in next commit

File: docs/pages/define/connectiontypes/db.bigquery.md (hunk 6 above)
Comment (peterbillen) [reply docs/pages/define/connectiontypes/db.bigquery.md:44]: changed to authMode with options: service, user, token, default

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/DefaultBigqueryDatabaseConnection.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+@Data
+@ToString(callSuper = true)
+@EqualsAndHashCode(callSuper = true)
+public class DefaultBigqueryDatabaseConnection extends BigqueryDatabaseConnection {
+
+    public DefaultBigqueryDatabaseConnection(String hostName, int portNumber, String project) {
+        super(getConnectionUrl(hostName, portNumber, project));
+    }
+
+    private static String getConnectionUrl(String hostName, int portNumber, String project) {
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/DefaultBigqueryDatabaseConnection.java:17]: why is this method static?

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/DefaultBigqueryDatabaseConnection.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        StringBuilder connectionUrl = new StringBuilder();
+        connectionUrl.append("jdbc:bigquery://");
+        connectionUrl.append(hostName);
+        if (portNumber > 0) {
+            connectionUrl.append(":");
+            connectionUrl.append(portNumber);
+        }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/DefaultBigqueryDatabaseConnection.java:24]: port is listed as a mandatory parameter in 'BigqueryDatabaseService'. When would the port be smaller than zero?

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/ServiceBigqueryDatabaseConnection.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    private static String getConnectionUrl(String hostName, int portNumber, String project, String serviceAccount, String keyPath) {
+        StringBuilder connectionUrl = new StringBuilder();
+        connectionUrl.append("jdbc:bigquery://");
+        connectionUrl.append(hostName);
+        if (portNumber > 0) {
+            connectionUrl.append(":");
+            connectionUrl.append(portNumber);
+        }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/ServiceBigqueryDatabaseConnection.java:24]: port is listed as a mandatory parameter in 'BigqueryDatabaseService'. When would the port be smaller than zero?

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/TokenBigqueryDatabaseConnection.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+            connectionUrl.append(refreshToken);
+        }
+        if (!clientId.isEmpty()) {
+            connectionUrl.append(";OAuthClientId=");
+            connectionUrl.append(clientId);
+        }
+        if (!clientSecret.isEmpty()) {
+            connectionUrl.append(";OAuthClientSecret=");
+            connectionUrl.append(clientSecret);
+        }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/TokenBigqueryDatabaseConnection.java:44]: the parameters are listed as a mandatory parameter in 'BigqueryDatabaseService'. Could you make them Optional and check if the value is null instead

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/UserBigqueryDatabaseConnection.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    private static String getConnectionUrl(String hostName, int portNumber, String project) {
+        StringBuilder connectionUrl = new StringBuilder();
+        connectionUrl.append("jdbc:bigquery://");
+        connectionUrl.append(hostName);
+        if (portNumber > 0) {
+            connectionUrl.append(":");
+            connectionUrl.append(portNumber);
+        }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/UserBigqueryDatabaseConnection.java:24]: port is listed as a mandatory parameter in 'BigqueryDatabaseService'. When would the port be smaller than zero?

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/BigqueryDatabaseService.java [hunk 12]
```diff
@@ -0,0 +1,166 @@
+package io.metadew.iesi.connection.database.bigquery;
+
+import io.metadew.iesi.connection.database.DatabaseHandler;
+import io.metadew.iesi.connection.database.ISchemaDatabaseService;
+import io.metadew.iesi.connection.database.SchemaDatabaseService;
+import io.metadew.iesi.connection.database.connection.DatabaseConnectionHandler;
+import io.metadew.iesi.metadata.definition.MetadataField;
+import io.metadew.iesi.metadata.definition.connection.Connection;
+
+public class BigqueryDatabaseService extends SchemaDatabaseService<BigqueryDatabase> implements ISchemaDatabaseService<BigqueryDatabase> {
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/BigqueryDatabaseService.java:10]: Does Bigquery use the limit X offset Y format for paginated retrieval of rows? If not, you should override 'public String refactorLimitAndOffset(String query)'. See io.metadew.iesi.connection.database.oracle.OracleDatabaseConnectionService for example

Comment (peterbillen) [reply core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/BigqueryDatabaseService.java:10]: Bigquery uses limit x offset y syntax -- see https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax for more information

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/DefaultBigqueryDatabaseConnection.java (hunk 7 above)
Comment (peterbillen) [reply core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/DefaultBigqueryDatabaseConnection.java:17]: Needs to be static since it is called before supertype constructor has been called

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/DefaultBigqueryDatabaseConnection.java (hunk 8 above)
Comment (peterbillen) [reply core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/DefaultBigqueryDatabaseConnection.java:24]: Resolved in next commit

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/TokenBigqueryDatabaseConnection.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    private static String getConnectionUrl(String hostName, int portNumber, String project, String accessToken, String refreshToken, String clientId, String clientSecret) {
+        StringBuilder connectionUrl = new StringBuilder();
+        connectionUrl.append("jdbc:bigquery://");
+        connectionUrl.append(hostName);
+        if (portNumber > 0) {
+            connectionUrl.append(":");
+            connectionUrl.append(portNumber);
+        }
```
[the remark this answers is not in the document]
Comment (peterbillen) [reply core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/TokenBigqueryDatabaseConnection.java:24]: Resolved in next commit

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/UserBigqueryDatabaseConnection.java (hunk 11 above)
Comment (peterbillen) [reply core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/UserBigqueryDatabaseConnection.java:24]: Resolved in next commit

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/TokenBigqueryDatabaseConnection.java (hunk 10 above)
Comment (peterbillen) [reply core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/TokenBigqueryDatabaseConnection.java:44]: Resolved in next commit

File: core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/DefaultBigqueryDatabaseConnection.java (hunk 7 above)
Comment (peterbillen) [reply core/java/iesi-core/src/main/java/io/metadew/iesi/connection/database/bigquery/DefaultBigqueryDatabaseConnection.java]: @BerrevoetsRobbe  Have a look, but I think this could be kept as such

**Review by BerrevoetsRobbe — CHANGES_REQUESTED** [review]

File: core/conf/application-connection-types.yml [hunk 14]
```diff
@@ -24,6 +78,7 @@ iesi:
             type: boolean
             mandatory: true
             encrypted: false
+>>>>>>> develop
```
[the remark this answers is not in the document]
Comment (peterbillen) [reply core/conf/application-connection-types.yml]: indeed - didn't notice, is corrected with next commit

**Review by BerrevoetsRobbe — APPROVED** [review]
